### PR TITLE
feat(wallet): add security settings

### PR DESCRIPTION
**Pull request by doums**
# Description

Closes: 
- #2275
- #3285
- #3275

### CODE REVIEW

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 1]
```diff
@@ -170,12 +192,27 @@ fn store_login_with_multiple_accounts_at_file(
...
+) -> Result<(), BackendError> {
+    let mut stored_wallet = match load_existing_wallet_at_file(filepath) {
+        Err(BackendError::WalletFileNotFound) => StoredWallet::default(),
+        result => result?,
+    };
 
-    stored_wallet.add_encrypted_login(new_encrypted_login)?;
+    archive_wallet_file()?;
+    let new_login = get_new_encrypted_login(mnemonic, hd_path, id, password)?;
+    stored_wallet.replace_encrypted_login(new_login)?;
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: Maybe I'm missing it, but what happens in the case for multiple logins? Will all passwords be updated? Should they?

Also, would be be good with some unit tests for these cases

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 2]
```diff
@@ -138,18 +138,40 @@ fn store_login_at_file(
     write_to_file(filepath, &stored_wallet)
 }
 
-pub(crate) fn store_login_with_multiple_accounts(
+/// Store the login with multiple accounts support
+///
+/// Set `overwrite` to true to overwrite the wallet file in case of password update
+pub(crate) fn save_login(
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: Imho, with overwrite = false you are effectively appending a login to a file, with overwrite = true you're not appending, but resaving existing logins already in the file. So it feels like the latter should be a separate function. Basically the first is a `.append()` and the latter is a `.for_each(mutate)`.

```
fn store_login_with_multiple_accounts()

fn re_store_all_logins()
```

**Review by octol — CHANGES_REQUESTED** [review]

File: nym-wallet/src-tauri/src/wallet_storage/account_data.rs [hunk 3]
```diff
@@ -108,6 +108,20 @@ impl StoredWallet {
         self.get_encrypted_login(id)?.decrypt_struct(password)
     }
 
+    pub fn reencrypt_login(
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/account_data.rs]: Looks good! I take it wasn't possible to call `decrypt_login` due to the lack of `mut`?

IMHO let's also do a `reencrypt_all`

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 4]
```diff
@@ -138,7 +138,8 @@ fn store_login_at_file(
     write_to_file(filepath, &stored_wallet)
 }
 
-pub(crate) fn store_login_with_multiple_accounts(
+/// Store the login with multiple accounts support
+pub(crate) fn save_encrypted_login(
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: "save" is fine, but I think in all the other functions "store" is used? If that's the case lets stay consistent

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 5]
```diff
@@ -152,6 +153,38 @@ pub(crate) fn store_login_with_multiple_accounts(
...
+    let mut stored_wallet = match load_existing_wallet_at_file(&filepath) {
+        Err(BackendError::WalletFileNotFound) => StoredWallet::default(),
+        result => result?,
+    };
+
+    stored_wallet.reencrypt_login(&id, current_password, new_password)?;
+    write_to_file(&filepath, &stored_wallet)
+}
+
+fn get_new_encrypted_login(
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: Can probably skip the `get_`

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 6]
```diff
@@ -152,6 +153,38 @@ pub(crate) fn store_login_with_multiple_accounts(
     store_login_with_multiple_accounts_at_file(&filepath, mnemonic, hd_path, id, password)
 }
 
+/// Update a login with multiple accounts support
+pub(crate) fn update_encrypted_login(
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: IMHO I think we should just reencrypt all logins, if nothing else just to keep the invariant that all logins have the same password

File: nym-wallet/src-tauri/src/operations/mixnet/account.rs [hunk 7]
```diff
@@ -304,7 +304,19 @@ pub fn create_password(mnemonic: Mnemonic, password: UserPassword) -> Result<(),
...
+
+#[tauri::command]
+pub fn update_password(
+    current_password: UserPassword,
+    new_password: UserPassword,
+) -> Result<(), BackendError> {
+    log::info!("Updating password");
+
+    // Currently we only support a single, default, login id in the wallet
+    let login_id = wallet_storage::LoginId::new(DEFAULT_LOGIN_ID.to_string());
```
Comment (octol) [inline nym-wallet/src-tauri/src/operations/mixnet/account.rs]: Just skip this an re-encrypt all logins (sorry duplicate comment)

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 8]
```diff
@@ -152,6 +153,48 @@ pub(crate) fn store_login_with_multiple_accounts(
     store_login_with_multiple_accounts_at_file(&filepath, mnemonic, hd_path, id, password)
 }
 
+/// Update all logins with multiple accounts support
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs:156]: You update all logins, whether each of them contain multiple accounts or not. The current wording makes it sounds like you're updating only the ones with multiple accounts support

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 9]
```diff
@@ -152,6 +153,48 @@ pub(crate) fn store_login_with_multiple_accounts(
...
+pub(crate) fn update_encrypted_logins(
+    current_password: &UserPassword,
+    new_password: &UserPassword,
+) -> Result<(), BackendError> {
+    // make sure the entire directory structure exists
+    let store_dir = get_storage_directory()?;
+    create_dir_all(&store_dir)?;
+    let filepath = store_dir.join(WALLET_INFO_FILENAME);
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs:162]: Just check if it exists? If you try to update something that doesn't exists then that seems like an error?

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 10]
```diff
@@ -633,6 +672,108 @@ mod tests {
...
+        update_encrypted_logins_at_file(&wallet_file, &password, &password).unwrap();
+
+        let loaded_login = load_existing_login_at_file(&wallet_file, &id1, &password).unwrap();
+        let acc = loaded_login.as_mnemonic_account().unwrap();
+        assert_eq!(&account1, acc.mnemonic());
+        assert_eq!(&hd_path, acc.hd_path());
+    }
+
+    #[test]
+    #[should_panic]
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: Avoid panic if you can, instead check that you get the correct error type

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 11]
```diff
@@ -633,6 +672,108 @@ mod tests {
...
+        store_login_at_file(
+            &wallet_file,
+            account1.clone(),
+            hd_path.clone(),
+            id1.clone(),
+            &password,
+        )
+        .unwrap();
+
+        update_encrypted_logins_at_file(&wallet_file, &wrong_password, &new_password).unwrap();
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: If this line is expected to panic, just unwrap_err and assert on the error type

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 12]
```diff
@@ -633,6 +672,108 @@ mod tests {
...
+    }
+
+    #[test]
+    #[should_panic]
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: Same, remove pls

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 13]
```diff
@@ -813,6 +1035,58 @@ mod tests {
...
+        let account1 = Mnemonic::generate(24).unwrap();
+        let account2 = Mnemonic::generate(24).unwrap();
+        let cosmos_hd_path: DerivationPath = COSMOS_DERIVATION_PATH.parse().unwrap();
+        let different_hd_path: DerivationPath = "m".parse().unwrap();
+        let password = UserPassword::new("password".to_string());
+        let new_password = UserPassword::new("new_password".to_string());
+        let id1 = LoginId::new("first".to_string());
+        let id2 = LoginId::new("second".to_string());
+
+        // Store the first account
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: "Store the first login with an account"

File: nym-wallet/src-tauri/src/wallet_storage/mod.rs [hunk 14]
```diff
@@ -813,6 +1035,58 @@ mod tests {
...
+            &password,
+        )
+        .unwrap();
+
+        let login = load_existing_login_at_file(&wallet, &id1, &password).unwrap();
+        let acc = login.as_mnemonic_account().unwrap();
+        assert_eq!(&account1, acc.mnemonic());
+        assert_eq!(&cosmos_hd_path, acc.hd_path());
+
+        // Add an extra account
```
Comment (octol) [inline nym-wallet/src-tauri/src/wallet_storage/mod.rs]: "Store a second login, also with an account"

**Review by octol — COMMENTED** [review]
Rust parts approved

**Review by octol — APPROVED** [review]

**Review by gala1234 — APPROVED** [review]
Looks good for me

**Pull request merged**